standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/properties/CachingPropertyStore.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.metastore.properties;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+import java.util.UUID;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ * A property map store using a pull-thru cache.
+ * Before a map is returned, a check against the stored corresponding digest is performed to avoid
+ * using stale data.
+ */
+public class CachingPropertyStore extends PropertyStore {
+  private static final int MAP_CACHE = 64;
+  protected  SoftCache<String, PropertyMap> maps;
+  protected PropertyStore store;
+  public CachingPropertyStore(PropertyStore wrap) {
+    this(wrap, MAP_CACHE);
+  }
+
+  public CachingPropertyStore(PropertyStore wrap, int capacity) {
+    store = wrap;
+    maps = new SoftCache<>(capacity, false);
+  }
+  public void clearCache() {
+    maps.clear();
+  }
+
+  @Override public UUID fetchDigest(String mapKey) {
+    return store.fetchDigest(mapKey);
+  }
+
+  @Override
+  public Map<String, UUID> selectDigest(String keyPrefix, Predicate<String> keyFilter) {
+    return store.selectDigest(keyPrefix, keyFilter);
+  }
+
+  @Override
+  public PropertyMap fetchProperties(final String mapKey, final Function<String, PropertySchema> getSchema) {
+    synchronized(this) {
+      PropertyMap map = maps.compute(mapKey, mapsCompute(mapKey, getSchema));
+      // we always return a copy of the properties in the cache
+      return map != null? map.copy() : null;
+    }
+  }
+
+  BiFunction<String, PropertyMap, PropertyMap> mapsCompute(String string, Function<String, PropertySchema> getSchema) {
```

Comment:
Should the `string ` argument be changed into something more meaningful?

Reply:
The map key is a string; it is generic by design since the PropertyManager is responsible for encoding/decoding whatever semantic is needed (schema resolution included). I tried with a class storing the various segments but it added lots of complications and constraints for very little abstraction gains (if any).